concurrent/src/main/scala/scalaz/concurrent/Task.scala

```diff
@@ -181,10 +270,28 @@ class Task[+A](val get: Future[Throwable \/ A]) {
    * A `Task` which returns a `TimeoutException` after `timeoutInMillis`,
    * and attempts to cancel the running computation.
    */
-  def timed(timeoutInMillis: Long)(implicit scheduler:ScheduledExecutorService): Task[A] =
-    new Task(get.timed(timeoutInMillis).map(_.join))
+  def timed(timeoutInMillis: Long)(implicit scheduler: ScheduledExecutorService): Task[A] =
+    new Task(Future.async[Throwable \/ A] { cb =>
+      val cancel = new AtomicBoolean(false)
+      val done = new AtomicBoolean(false)
+      scheduler.schedule(new Runnable {
+        def run() {
+          if (done.compareAndSet(false,true)) {
+            cancel.set(true)
+          }
+        }
+      }
+      , timeoutInMillis, TimeUnit.MILLISECONDS)
+      unsafePerformListenInterruptibly(a => Trampoline.done({
+        if(done.compareAndSet(false, true)) 
+          cb(a)
+        else
+          cb(-\/(new TimeoutException(s"Timed out after $timeoutInMillis milliseconds")))
+        }), cancel)
```

Comment:
This fixes the interactions between consecutive `timed` calls, though it still breaks monadic associativity.